language/bytecode-verifier/src/dependencies.rs

```diff
@@ -109,289 +76,266 @@ impl<'a> DependencyChecker<'a> {
                 }
                 let func_handle = module.function_handle_at(func_def.function);
                 let func_name = module.identifier_at(func_handle.name);
-                self.func_id_to_handle_map
+                checker
+                    .func_id_to_handle_map
                     .insert((module_id.clone(), func_name.to_owned()), func_def.function);
             }
         }
+        checker
     }
+}
 
-    fn verify_imported_modules(
-        &self,
-        module_handles: &[ModuleHandle],
-        self_module: Option<ModuleHandleIndex>,
-    ) -> PartialVMResult<()> {
-        for (idx, module_handle) in module_handles.iter().enumerate() {
-            let module_id = self.resolver.module_id_for_handle(module_handle);
-            if Some(ModuleHandleIndex(idx as u16)) != self_module
-                && !self.dependency_map.contains_key(&module_id)
-            {
-                return Err(verification_error(
-                    StatusCode::MISSING_DEPENDENCY,
-                    IndexKind::ModuleHandle,
-                    idx as TableIndex,
-                ));
-            }
+pub fn verify_module<'a>(
+    module: &CompiledModule,
+    dependencies: impl IntoIterator<Item = &'a CompiledModule>,
+) -> VMResult<()> {
+    verify_module_impl(module, dependencies)
+        .map_err(|e| e.finish(Location::Module(module.self_id())))
+}
+
+fn verify_module_impl<'a>(
+    module: &CompiledModule,
+    dependencies: impl IntoIterator<Item = &'a CompiledModule>,
+) -> PartialVMResult<()> {
+    let checker = &DependencyChecker::module(module, dependencies);
+
+    verify_imported_modules(checker)?;
+    verify_imported_structs(checker)?;
+    verify_imported_functions(checker)
```

Comment:
Would it be better to move these `verify_imported_*` functions into `DependencyChecker`? So we can do
`checker.verify_imported_modules()?;` ...

Reply:
No. These aren't operations belonging to checker. The checker is just a context under which these checks are performed. 

So there is an intentional API separation there (this is why I hate the checker code that exists today, all of the logic is all bundled together and there is no clear split between context of the check and the checks themeslves)

I'll probably rename all of these in this file to maybe alleviate confusion